cmd/kcp-catalog/bind/catalogentry/bind.go

```diff
@@ -0,0 +1,214 @@
+}
+
+// Complete ensures all fields are initialized.
+func (b *BindOptions) Complete(args []string) error {
+	if err := b.Options.Complete(); err != nil {
+		return err
+	}
+
+	if len(args) > 0 {
+		b.CatalogEntryRef = args[0]
+	}
+	return nil
+}
+
+// Validate validates the BindOptions are complete and usable.
+func (b *BindOptions) Validate() error {
+	if b.CatalogEntryRef == "" {
+		return errors.New("`root:ws:catalogentry_object` reference to bind is required as an argument")
+	}
+
+	if !strings.HasPrefix(b.CatalogEntryRef, "root") || !logicalcluster.New(b.CatalogEntryRef).IsValid() {
+		return fmt.Errorf("fully qualified reference to workspace where catalog entry exists is required. The format is `root:<ws>:<catalogentry>`")
+	}
+
+	return b.Options.Validate()
+}
+
+// Run creates an apibinding for the user.
+func (b *BindOptions) Run(ctx context.Context) error {
+	config, err := b.ClientConfig.ClientConfig()
+	if err != nil {
+		return err
+	}
+
+	baseURL, currentClusterName, err := pluginhelpers.ParseClusterURL(config.Host)
+	if err != nil {
+		return err
+	}
+
+	// get the base config, which is needed for creation of clients.
+	baseConfig, err := ctrlcfg.GetConfigWithContext("base")
+	if err != nil {
+		return fmt.Errorf("unable to get base config %v", err)
+	}
+
+	// get the entry referenced in the command to which the user wants to bind.
+	path, entryName := logicalcluster.New(b.CatalogEntryRef).Split()
+	cfg := rest.CopyConfig(config)
+	cfg.Host = baseURL.String()
+	client, err := newClient(cfg, path)
+	if err != nil {
+		return err
+	}
+
+	entry := catalogv1alpha1.CatalogEntry{}
+	err = client.Get(ctx, types.NamespacedName{Name: entryName}, &entry)
+	if err != nil {
+		return fmt.Errorf("cannot find the catalog entry %q referenced in the command in the workspace %q", entryName, path)
+	}
+
+	kcpClient, err := newClient(baseConfig, currentClusterName)
+	if err != nil {
+		return err
+	}
+
+	allErrors := []error{}
+
+	apiBindings := []apisv1alpha1.APIBinding{}
+	for _, ref := range entry.Spec.Exports {
+		// check if ref is valid. Skip if invalid by logging error.
+		if ref.Workspace.Path == "" || ref.Workspace.ExportName == "" {
+			if _, err := fmt.Fprintf(b.Out, "invalid reference %q/%q", ref.Workspace.Path, ref.Workspace.ExportName); err != nil {
+				allErrors = append(allErrors, err)
+			}
+			continue
+		}
+
+		apiBinding := &apisv1alpha1.APIBinding{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: fmt.Sprintf("%s-%s", ref.Workspace.ExportName, utilrand.String(10)),
+			},
+			Spec: apisv1alpha1.APIBindingSpec{
+				Reference: ref,
+			},
+		}
+
+		apiBindings = append(apiBindings, *apiBinding)
+	}
+
+	// Create bindings to the target workspace
+	for _, binding := range apiBindings {
+		err := kcpClient.Create(ctx, &binding)
+		if err != nil {
+			// If an APIBinding already exists, intentionally not updating it since we would not like reset AcceptablePermissionClaims.
```

Comment:
Now that you're using a partially random name, you won't get an "already exists" error. The binding creation should succeed, but when the controllers reconcile it, they may find conflicts (you'll see them in status.conditions).

Reply:
I wonder if we should check the existing APIBindings in the workspace and see if any of them carry the same ExportReferences. If that is the case, then simply not creating a new APIBinding for that particular ExportReference.

Reply:
@dinhxuanvu I intentionally didn't add that check because the permission claims in the already created binding might be different. I'm not sure if just updating permission claims of an existing claim is something which we should do. There are chances wherein the permission claims in exports ref could be different than the one existing in apibinidng present in cluster. For now, I'm just overwriting the permission claims in the existing apibinding and doing an update. Not sure if that is the approach we want.

Reply:
I think given the fact the new APIBinding will fail due to what Andy said early, it is not a good UX to create a binding that won't even do anything but getting into a failed state. If the existing binding is missing some permissions, that can be resolved without having to delete the extra binding that we shouldn't create in the first place. Either way, would like to hear what Andy thinks about this.

Reply:
I would recommend something like this:
```
$ k catalog bind catalogentry root:catalog:awesomeness
Binding for "awesomeness" already exists but permission claims are different:

<show differences>

Rerun with --update to update the existing binding's permission claims.
```

But that can be in a follow-up.